float32 reward
int32 length